### ISSUE TITLE
SILE is single threaded and hence slow

### ISSUE THREAD

**Issue by alerque** [sile-typesetter/sile#884]
I have a new toy. Quarantine rules have kept me away from my office (and office computer). The rig I was making due with at home was a 12 year old build that has finally bit the dust. I just replaced it with a new rig: AMD Ryzen 5 3600, 64 Gigs of RAM, and some NVMe. The difference is like waking up on a new planet.

One side effect of having 12 threads and endless RAM to throw at problems instead of 2 & a pinch is that it's now painfully obvious what operations make use of threading and which are putzing along in the slow lane. I can't even imagine what it would be like if I'd shelled out for a Ryzen Threadripper! With silicon Mhz not going up that much but more and more parallel power to throw around in even modest budget consumer equipment I think we need to re-evaluate SILE's main loop.

Obviously most of what it does is linear and isn't easily short circuited. However I think we probably can take some chucks out of the time it takes to render a long document if we think about it.

**Comment by alerque** [comment]
One thing I think we could do right away is actually in relation to #230. Instead of resolving this by isolating the main rendering loop and running it 2 or 3 times, we may be able to launch 3 threads instead. A couple things would have to happen off the top of my head:

* The AST would need to be fully parsed first, and some sort of indexing added to it.
* Each thread would need to need to check in with the one before it and make sure it never got ahead of it in the AST.
* Info tables like TOC data would need to be kept and passed in memory instead of being dumped to a file only at the end of a run.
* An extra thread dependency tracker would be needed so that the 3rd thread understood it wasn't allowed past the TOC until the 1st thread was finished.

This wouldn't end up being 3 times faster, but it might end up being a shade over twice as fast to get 3 rendering passes done.

**Comment by simoncozens** [comment]
I could imagine maybe doing a pipeline where the AST parser hands nodes to the line builder which hands them to the page builder which hands them to the output bit. I have no idea how to approach multithreading in Lua, though. Quick Googling looks like https://github.com/LuaLanes/lanes is the way to go, but coordination/communication between threads looks messy.

**Comment by alerque** [comment]
I messed with lanes a bit for another project and it is messy, but also the least messy I've seen in Lua. None of it holds a candle to how nice thread handling is in, say, Rust.

**Comment by alerque** [comment]
Similar to the hand-off strategy [mentioned](https://github.com/sile-typesetter/sile/issues/884#issuecomment-641477781), but on a bit more aggressive level, there are a lot of loops that spend quite a bit of time re-iterating over the same set of nodes looking for the least-bad break points and other math like that. Some of these loops go over the same ground a lot of times. If we could break out the data make decent functions out of the loop actions we can probable run several of these at a time and compare values in a separate loop. Even if we end up wasting some CPU time with results we won't use, calculating N+1 sets of data in parallel will probably be a _lot_ faster that waiting to for every iteration in serial.

**Comment by simoncozens** [comment]
Yeah, the line-breaking code is a hotspot. But as it's the Knuth algorithm, it really ought to be as optimized as you're going to get as far as the algorithm goes. It may be that there's just a lot of overhead with shuffling objects around.

In [newbreak](https://github.com/simoncozens/newbreak/blob/master/newbreak.ts) I use a deliberately dumb algorithm and rely on memoization to speed it up considerably. I'd like to do a newbreak implementation for SILE, preferably in a compiled language to avoid the overhead. I tried writing one in C but there were segfaults everywhere... I've written a Rust implementation but am having problems working out how to handle the C API - the "linebreak" object has lifetime bounds and I don't know how to hand an object like that back to C.

**Comment by alerque** [comment]
I don't think passing back to C is the way to go. I've been messing around with Rust & Lua interfaces and think directly going between them using mlua (or one of it's variants) would be preferable. Lifetime management is one reason. Better handling of type checking/casting is another. We can fairly easily write pure Rust modules that can be loaded directly into Lua (basically using the C api, but transparently skipping any C). I was already thinking of this as a way to swap out some functions for speed while still allowing them to be implemented in Lua for extensibility.

**Comment by simoncozens** [comment]
Either way you have to get a lifetimed struct across the FFI bridge into a world that doesn't know about lifetimes. I'm sure there is a way to do it, I just don't know what it is.

Meanwhile I've also got a C++ implementation that seems to work...
